### ISSUE TITLE
Error with fp8_e4m3fn_scaled_fast when CFG > 1

### ISSUE THREAD

**Issue by tuolaku** [kijai/ComfyUI-WanVideoWrapper#1204]
I’ve been running fp8_e4m3fn_scaled_fast without any issues as long as CFG is set to 1.
Today, while experimenting with the “WanVideoScheduler” node, I found that raising CFG above 1 triggers the following error:

# ComfyUI Error Report
## Error Details
- **Node ID:** 139
- **Node Type:** WanVideoSampler
- **Exception Type:** RuntimeError
- **Exception Message:** Invalid scaling configuration.
- For TensorWise scaling, a and b should be float8, scales should be float and singletons.
- For RowWise scaling, a and b should be float8, scales should be float, scale_a should be (114048, 1) and scale_b should be (1, 5120), and both should be contiguous.
- For BlockWise 1x128 scaling, a and b should be float8, scales should be float, scale_a should be (114048, 40) and scale_b should be (40, 5120), and both should be outer-dim-major.
- For BlockWise 128x128 scaling, a and b should be float8, scales should be float, scale_a should be (891, 40) and scale_b should be (40, 40), and both should be near-inner-dim-major (with 16-byte aligned strides).
- For Blockwise 1x32 scaling, a and b should be float8, scales should be float8_e8m0fnu, scale_a should have 18247680 elements and scale_b should have 819200 elements, and both should be contiguous.
- For Blockwise 1x16 scaling, a and b should be float4 (packed 2x), scales should be float8_e4m3fn, scale_a should have 72990720 elements and scale_b should have 3276800 elements, and both should be contiguous.
Got a.dtype()=Float8_e4m3fn, scale_a.dtype()=Float, scale_a.size()=[], scale_a.stride()=[], b.dtype()=Float8_e4m3fn, scale_b.dtype()=Half, scale_b.size()=[1] and scale_b.stride()=[1]

## Stack Trace
```
  File "D:\sd-comfyui-aki\execution.py", line 496, in execute
    output_data, output_ui, has_subgraph, has_pending_tasks = await get_output_data(prompt_id, unique_id, obj, input_data_all, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, hidden_inputs=hidden_inputs)
                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "D:\sd-comfyui-aki\execution.py", line 315, in get_output_data
    return_values = await _async_map_node_over_list(prompt_id, unique_id, obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, hidden_inputs=hidden_inputs)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "D:\sd-comfyui-aki\custom_nodes\ComfyUI-Lora-Manager\py\metadata_collector\metadata_hook.py", line 165, in async_map_node_over_list_with_metadata
    results = await original_map_node_over_list(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<2 lines>...
    )
    ^

  File "D:\sd-comfyui-aki\execution.py", line 289, in _async_map_node_over_list
    await process_inputs(input_dict, i)

  File "D:\sd-comfyui-aki\execution.py", line 277, in process_inputs
    result = f(**inputs)

  File "D:\sd-comfyui-aki\custom_nodes\ComfyUI-WanVideoWrapper\nodes.py", line 4056, in process
    raise e

  File "D:\sd-comfyui-aki\custom_nodes\ComfyUI-WanVideoWrapper\nodes.py", line 3952, in process
    noise_pred, self.cache_state = predict_with_cfg(
                                   ~~~~~~~~~~~~~~~~^
        latent_model_input,
        ^^^^^^^^^^^^^^^^^^^
    ...<2 lines>...
        timestep, idx, image_cond, clip_fea, control_latents, vace_data, unianim_data, audio_proj, control_camera_latents, add_cond,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        cache_state=self.cache_state, fantasy_portrait_input=fantasy_portrait_input, mtv_motion_tokens=mtv_motion_tokens, s2v_audio_input=s2v_audio_input)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "D:\sd-comfyui-aki\custom_nodes\ComfyUI-WanVideoWrapper\nodes.py", line 2909, in predict_with_cfg
    raise e

  File "D:\sd-comfyui-aki\custom_nodes\ComfyUI-WanVideoWrapper\nodes.py", line 2810, in predict_with_cfg
    noise_pred_cond, cache_state_cond = transformer(
                                        ~~~~~~~~~~~^
        [z], context=positive_embeds, y=[image_cond_input] if image_cond_input is not None else None,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
        **base_params
        ^^^^^^^^^^^^^
    )
    ^

  File "D:\sd-comfyui-aki\.ext\Lib\site-packages\torch\nn\modules\module.py", line 1775, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^

  File "D:\sd-comfyui-aki\.ext\Lib\site-packages\torch\nn\modules\module.py", line 1786, in _call_impl
    return forward_call(*args, **kwargs)

  File "D:\sd-comfyui-aki\custom_nodes\ComfyUI-WanVideoWrapper\wanvideo\modules\model.py", line 2485, in forward
    x, x_ip = block(x, x_ip=x_ip, **kwargs) #run block
              ~~~~~^^^^^^^^^^^^^^^^^^^^^^^^

  File "D:\sd-comfyui-aki\.ext\Lib\site-packages\torch\_dynamo\eval_frame.py", line 414, in __call__
    return super().__call__(*args, **kwargs)
           ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^

  File "D:\sd-comfyui-aki\.ext\Lib\site-packages\torch\nn\modules\module.py", line 1775, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^

  File "D:\sd-comfyui-aki\.ext\Lib\site-packages\torch\nn\modules\module.py", line 1786, in _call_impl
    return forward_call(*args, **kwargs)

  File "D:\sd-comfyui-aki\.ext\Lib\site-packages\torch\_dynamo\eval_frame.py", line 832, in compile_wrapper
    return fn(*args, **kwargs)

  File "D:\sd-comfyui-aki\.ext\Lib\site-packages\torch\nn\modules\module.py", line 1775, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^

  File "D:\sd-comfyui-aki\.ext\Lib\site-packages\torch\nn\modules\module.py", line 1786, in _call_impl
    return forward_call(*args, **kwargs)

  File "D:\sd-comfyui-aki\custom_nodes\ComfyUI-WanVideoWrapper\wanvideo\modules\model.py", line 965, in forward
    q, k, v = self.self_attn.qkv_fn(input_x)
              ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^

  File "D:\sd-comfyui-aki\custom_nodes\ComfyUI-WanVideoWrapper\wanvideo\modules\model.py", line 390, in qkv_fn
    q = self.norm_q(self.q(x)).view(b, s, n, d)
                    ~~~~~~^^^

  File "D:\sd-comfyui-aki\.ext\Lib\site-packages\torch\nn\modules\module.py", line 1775, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^

  File "D:\sd-comfyui-aki\.ext\Lib\site-packages\torch\nn\modules\module.py", line 1786, in _call_impl
    return forward_call(*args, **kwargs)

  File "D:\sd-comfyui-aki\custom_nodes\ComfyUI-WanVideoWrapper\fp8_optimization.py", line 63, in <lambda>
    setattr(submodule, "forward", lambda input, m=submodule: fp8_linear_forward(m, base_dtype, input))
                                                             ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^

  File "D:\sd-comfyui-aki\custom_nodes\ComfyUI-WanVideoWrapper\fp8_optimization.py", line 24, in fp8_linear_forward
    o = torch._scaled_mm(inn, cls.weight.t(), out_dtype=base_dtype, bias=bias, scale_a=scale_input, scale_b=scale_weight)

**Comment by tuolaku** [comment]
I suspect the problem is related to the way the scale is directly taken from the input tensor inside the fp8_linear_forward() function.

**Comment by tuolaku** [comment]
Another problem: after adding the “WanVideoScheduler” node, the steps, start_steps, and stop_steps parameters in the low-noise part seem to be ignored.
For example, when the low-noise phase is supposed to run for 13 steps, I want it to abort after only 4 steps, so I set start_steps = 7 and stop_steps = 11.
Nevertheless, the scheduler still carries on for the full 13 low-noise steps and continues all the way to the final 20 steps.

**Comment by kijai** [comment]
I can't replicate the issue, fp8_fast works fine with cfg or not... which exact model are you using?

As to the WanVideoScheduler, the scheduler -output overrides every other setting in the sampler.

**Comment by tuolaku** [comment]
> 我无法复制这个问题，fp8_fast是否与 cfg 一起使用......您使用的确切型号是什么？
> 
> 对于 WanVideoScheduler，调度程序 -output 会覆盖采样器中的所有其他设置。

Here are the key screenshots for your reference:

<img width="2247" height="626" alt="Image" src="https://github.com/user-attachments/assets/82f1d9e7-8f62-4a2e-84b8-2be029fe83f7" />

<img width="1626" height="1206" alt="Image" src="https://github.com/user-attachments/assets/ddb9af2c-4327-45c8-9943-4b93405bc94b" />

When CFG = 1, the model only executes the “conditional” branch; it never runs the “unconditional” one.
As soon as CFG > 1, both branches must be evaluated. Inside WanVideo’s FP8 kernel, the two branches share the same quantization scales. The first branch quietly overwrites those scales in-place, converting them to Half/incorrect shapes. When the second branch arrives, torch._scaled_mm sees the corrupted scales and throws the validation error.
Code-path difference
CFG = 1
predict_with_cfg(...) is invoked once → a single transformer() call → FP8 scales are initialized once → no conflict.
CFG > 1
The negative (unconditional) branch runs first, then the positive (conditional) branch.
Both branches reuse the same FP8 linear object, and the scales are stored as run-time cached attributes of that object.
Why do the scales change?
Inside fp8_linear_forward(), to save memory, the scale is taken directly from the input tensor:

scale = input.abs().max()
Up-stream nodes (e.g., the sampler) may cast intermediate activations to Half before the second call.
max() keeps the dtype of its source tensor, so the second branch receives a Half-typed scale; feeding that into torch._scaled_mm violates the FP8 contract.

this is my preliminary analysis via Kimi—hope it helps.

**Comment by kijai** [comment]
Can you try without NAG?

**Comment by tuolaku** [comment]
> 你能在没有 NAG 的情况下尝试吗？

The error still occurs even with NAG disabled.

**Comment by tuolaku** [comment]
> 你能在没有 NAG 的情况下尝试吗？

Root cause
In the default Tensor-wise mode, torch._scaled_mm demands that
scale_a and scale_b be 0-D float32 tensors living on the CPU (shape () scalars)
They must not reside on the GPU
Your code, however, does
Python

scale_input = torch.ones((), device=input.device, dtype=torch.float32)
scale_weight = scale_weight.to(input.device)   # ← moved to GPU
placing the scale tensors on the GPU and triggering the RuntimeError.
Fix
Keep both scales on the CPU and leave the rest of the code unchanged:
Python

def fp8_linear_forward(cls, base_dtype, input):
    weight_dtype = cls.weight.dtype
    if weight_dtype in [torch.float8_e4m3fn, torch.float8_e5m2]:
        if len(input.shape) == 3:
            input_shape = input.shape

            # ---------- changes start ----------
            scale_weight = getattr(cls, 'scale_weight', None)
            if scale_weight is None:
                scale_weight = torch.ones((), dtype=torch.float32)   # stay on CPU
            else:
                scale_weight = scale_weight.to('cpu', dtype=torch.float32)  # force CPU + float32

            scale_input = torch.ones((), dtype=torch.float32)        # stay on CPU
            # ---------- changes end ----------

            input = torch.clamp(input, min=-448, max=448, out=input)
            inn = input.reshape(-1, input_shape[2]).to(torch.float8_e4m3fn).contiguous()

            bias = cls.bias.to(base_dtype) if cls.bias is not None else None

            o = torch._scaled_mm(inn, cls.weight.t(),
                                 out_dtype=base_dtype,
                                 bias=bias,
                                 scale_a=scale_input,
                                 scale_b=scale_weight)

            return o.reshape((-1, input_shape[1], cls.weight.shape[0]))
        else:
            return cls.original_forward(input.to(base_dtype))
    else:
        return cls.original_forward(input)
I'm not sure if this analysis is correct; it's for your reference only.

**Comment by tuolaku** [comment]
> 你能在没有 NAG 的情况下尝试吗？

Major finding: this error only occurs when the WanVideo Model Loader is not connected to any LoRA. As soon as a LoRA is plugged in, the workflow runs without throwing the RuntimeError.

Error:

<img width="2209" height="596" alt="Image" src="https://github.com/user-attachments/assets/91fd523a-1a3e-4f23-8d68-acb7e4b96929" />

the workflow runs:

<img width="2184" height="619" alt="Image" src="https://github.com/user-attachments/assets/cb3a0119-1b18-4644-bacf-e41b56b41dc0" />

**Comment by kijai** [comment]
> > 你能在没有 NAG 的情况下尝试吗？
> 
> Major finding: this error only occurs when the WanVideo Model Loader is not connected to any LoRA. As soon as a LoRA is plugged in, the workflow runs without throwing the RuntimeError.
> 
> Error:
> 
> <img alt="Image" width="2000" height="596" src="https://private-user-images.githubusercontent.com/193845733/486842444-91fd523a-1a3e-4f23-8d68-acb7e4b96929.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczNDY5MTcsIm5iZiI6MTc1NzM0NjYxNywicGF0aCI6Ii8xOTM4NDU3MzMvNDg2ODQyNDQ0LTkxZmQ1MjNhLTFhM2UtNGYyMy04ZDY4LWFjYjdlNGI5NjkyOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQxNTUwMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MjE0ODRlM2UzMzBjMjQ2YWIxYThiZjU3NGEyYTdkNjczNTVhYjlkMzEzMTg1YzE0Mzg1NGE5N2FlMDE4YmE4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.PROljuh1YgXC_VM3a3ViqYuTQtcxs21EgJn7VTkymbU">
> the workflow runs:
> 
> <img alt="Image" width="2000" height="619" src="https://private-user-images.githubusercontent.com/193845733/486842796-cb3a0119-1b18-4644-bacf-e41b56b41dc0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczNDY5MTcsIm5iZiI6MTc1NzM0NjYxNywicGF0aCI6Ii8xOTM4NDU3MzMvNDg2ODQyNzk2LWNiM2EwMTE5LTFiMTgtNDY0NC1iYWNmLWU0MWI1NmI0MWRjMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQxNTUwMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kOTc3NjQ0OGFkNWEyNzIxOTUyM2QyODgxZjJiMjI2NmJjYzkxMzdlZWFhNjk2ZGExZTk3YjJjODdjMTI2YTZiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.W2twszkfLF8DTFnC_jNq7DTQig4Nj6MfgqrORzMOXik">

Oh, okay, thanks, should be fixed now then.
